File: 39. Combination_Sum/solution.cpp
class Solution {
public:
    
    vector<vector<int>> ans;
    
    void recur(vector<int>& candidates, int target, vector<int> cur_vec, int cur_idx){
        
        if(target == 0){
            ans.push_back(cur_vec);
            return;
        }
            
        while(cur_idx < candidates.size() && target - candidates[cur_idx] >= 0){
            
            cur_vec.push_back(candidates[cur_idx]);
            
            recur(candidates, target - candidates[cur_idx], cur_vec, cur_idx);
            cur_idx++;
            
            cur_vec.pop_back();
        }
        
        
    }
    
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        
        sort(candidates.begin(), candidates.end());
        
        candidates.erase(unique(candidates.begin(), candidates.end()), candidates.end());
        
        vector<int> start;
        recur(candidates, target, start, 0);
        return ans;
        
    }
};

File: 39. Combination_Sum/solution.java
class Solution {
    
    List<List<Integer>> ans = new ArrayList<List<Integer>>();
    
    public void recur(int[] candidates, int target, ArrayList<Integer> cur_vec, int idx){
        
        if(target == 0){
            ArrayList<Integer> tmp = new ArrayList<Integer>(cur_vec);
            ans.add(tmp);
            return;
        }
        
        int i = idx;
        
        
        while(i < candidates.length && (target - candidates[i]) >= 0){
            
            cur_vec.add(candidates[i]);
            recur(candidates, target - candidates[i], cur_vec, i);
            i++;
            cur_vec.remove(cur_vec.size() - 1);
            
        }
        
        return;
        
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        Set<Integer> tmp_set = new TreeSet<Integer>();
        
        for(int i = 0 ; i < candidates.length ; i++)
            tmp_set.add(candidates[i]);

        
        int[] new_candidates = new int[tmp_set.size()];
        
        int i = 0;
        for(Integer cur_val : tmp_set){
            new_candidates[i] = cur_val;
            i++;
        }
        
        ArrayList<Integer> start = new ArrayList<Integer>();
        
        recur(new_candidates, target, start, 0);
        
        return ans;
        
    }
}